### ISSUE TITLE
Project status

### ISSUE THREAD

**Issue by wootguy** [wootguy/SevenKewp#8]
Most 3.0 era Sven Co-op maps should work as is. Many 4.x era maps also work but may need ripenting and audio format conversions (script coming later). See the current list of supported maps [here ](https://github.com/wootguy/SevenKewp/issues/7).

Some major changes you'll notice is that custom SC weapons are replaced with stock Half-Life weapons (uzi/saw/minigun/m16 -> mp5, sniper -> crossbow, deagle -> 357). Unimplmented monsters are also remapped (kingpin -> tor, stukabat -> alien controller). Some basic features are implented in plugins which haven't been added here yet (player unstucking, thirdperson camera, etc.).

Current priority is adding support for more maps. After that maybe AI reworking/fixes to handle multiple players better.

Bugs noted during public testing:
```
- hornets and shock beams are physical
- death animation freezes if respawn too fast
- reload anims for players
- not all monsters follow
- don't flinch monsters so much
- monsters just stare at you sometimes
- shock troopers not retaliating against pit drones (auspices)
- grunt/otis speech missing
- saw shoots in tiny bursts
- scripted_sentence doesnt work (convert to normal sounds?)
- grunts crouching to attack which obscures their view and they shoot their cover instead
- force players out of the way of scripted sequences (extended, c1a0c)
- add kingpin, stukabat, sqknest, rpg grunt
- garg pain sound spam
- babygarg pain sounds when dead
- skill cfg (megamonstermassacre)
- aslave sometimes does double damage but hard to reproduce (double animation event probably)
- func_tank queues up bullets for one super shot after you come out of cover
- grunt uses turret 50 feet away (hostage2a)
- shocktroopers gibbed themselves underwater (rhapsody2)
- monsters trouble routing up slope (extension babygarg)
- EF_NOINTERP doesn't work (sc_waterworks tunnel trigger getaway_chase)
- global sound replacement for mod (ub_megaman2 pit drone custom models)
- player kept score on rejoin?
- no sprite effect for crowbar/grapple pickup
- tor fliching(?) too much (megaman3 boss mp5 stunlock)
- trigger_push is stronger than in sven (shockraid_jungle jungle teleport at spawn)
- breakable with 0 health shows no health(?) (blackmesaepf pc)
- right click gauss should gib robos
- no blood from egon
- show hud what opens doors, what buttons do, etc. (random shitty mapper logic)
- shockroaches always spin around when trooper gibbed
- mosnters dont touch trigger_once on elevator (justcause02 ending)
- monsters can't walk over opened door (oberon14)
- custom muzzle flashes/events
- clear text menu on server join
- warn if no worldspawn entity
- can zoom in with xbox during cutscene
- bodyguard minigun needs hwgrunt ai
- starting server on infested2 causes UserMsg client disconnects, even though entity data is fine (https://github.com/ValveSoftware/halflife/issues/1894)
- AllocBlock full after 100 map changes or so (client engine problem)
- "Cache_Free: not allocated" server crash after ~285 map changes 
- too many gibs visible ent limit (sence_bdayfun_b2)
- "monster_flyer" (sc_verc_18)
- monsters don't turn around to fight (starcraft)
- friendly npc skins
- any explosion dmg should gib corpse
- tb_mesa2bgold ladder in vat area teleport you (and non-ripent version freezes client)
- no weapons assassins easily stun locked with egon or smth
- hwgrunt has trouble targetting things (daikon_warfare1)

low prio:
- scversion world keyvalue probably affects lots of ents in unknown ways

unreproducible bugs:
- weird bug where g-ara4 door didnt open. Door said it was open already in debug. Server running for a while
- gordonsci2 sequence never plays for button. He just stares at the button
- tb_fun1 PM_Move crash jumping across pendulums
- tb_mesa5gold pm_move crash sitting at gonome spawn trigger and E moving a bit
- tb_mesa6gold pm_move crash just sitting on rail on tram (Expression: pm_shared_initialized) (gcTextures set to smth insanely high in PM_FindTextureType)
```

**Comment by wootguy** [comment]
I've noticed that new players can't do most sven maps because they're poorly designed and give no clues on what to do. So, I'm also working on making a plugin that tells you what to do and where to go in every map, complete with pathfinding that accounts for what a player can/can't reach in the moment. This will speed up map testing, detect when a map is softlocked, and lay the groundwork for better pathfinding for bots and npcs in the future. A huge project on its own. This is also high priority because whenever I do add new maps, I often see players complaining:
```
X: this map sucks
X: what do we do
X: pls rtv
X: rtvvvvvvvvvvvvv
- X has left the game.
```
In sven this is less of a problem because there is usually a veteran there who's played every map 30 times. HL on the other hand got an influx of new players who are still learning how to crouch jump.

**Comment by wootguy** [comment]
I've added about 400 maps in one year. There are 1200+ left according to SCMapDB. At the current pace, every sven map should be working by 2028. Almost all sven entities and features have been ported so the rest of the maps should come much faster, but there's still a good amount of scripted maps and svengine maps which exceed HL limits.

The last big feature to add, which players will notice, is creating a custom client for the weapons that need client-side prediction (uzi, minigun, deagle, m16, sniper, saw). That's low priority because I doubt many will be willing to manually install something like that. I also can't _guarantee_ a custom client won't trigger a VAC ban, and there are players who already use a custom client like AG or BugFixedHL. The client I make would probably be used only by myself and maybe 3 other enthusiasts. A waste of time while I'm happy with the HL guns.

I'm interested to see if it's feasible to have every type of GoldSrc client play on the same server (HL, sven, counter-strike, opposing force, xash, etc.). It would be cool to play alongside counter-strike players who have their own guns and movement. That should fill up servers more as well. Something to try after the MVP of this mod is released. The tedious work of replicating sven has to be done first.

**Comment by wootguy** [comment]
Priorities switched early summer because it looked like the server was going to die if I kept focusing on map ports. It's gone from record lows to record highs after adding chat sounds, anime girl models, and a custom client (1400 -> 2000 [player hours](https://wootguy.github.io/sctracker/)). I'm savoring this and working more on fun/cool features because I don't have any more aces up my sleeve to attract players the next time this happens. 3 years is a long time to keep players interested while slowly chipping away at the map pool.

**Comment by wootguy** [comment]
I only added about 200 maps in 2025, but all major porting problems are now solved. The last one was a newly discovered [world leaves limit](https://github.com/ValveSoftware/halflife/issues/3843#issuecomment-2814383140). I spent a couple months updating bspguy, mostly to solve that. bspguy now has tools which have fixed every worldleaf overflow I've encountered (v6 coming soon).

Progress on new maps is slow because I'm picking harder ones which are more fun to work on. I still think I'm on track for a 2028 mod release with every sven map being playable. I worry most about grinding through the hundreds of low-effort maps and svengine maps that need re-merging/subdividing. Low difficulty mind numbing jobs.

If anyone wants to help out, there's a [porting guide](https://github.com/wootguy/SevenKewp/wiki/Mapping-&-Porting-Guide) here. It's not easy, but watching people play something you made is super fun (just don't read the chat...). Send me a port and we'll force people to test it on [my server](https://wootguy.github.io/hlcoop_web/). Shoutouts to Sneaky Ema and Wayne for contributing over 100 maps so far.

<details>
  <summary>Contributions</summary>
  
```
Wayne edits/merges/ports:
------------------
actual_hl
afrikakorps_hl
blow_up_hla
combined
coop_hl
craxnhax_hl
cryocell_hl
ectopit_hl
enc_hl
enc_hl
gemerald
mom_v1
sc_grinch_hl
sc_lexm_v1
sc_wilkes_hl
sectore_level_hl_1 sectore_level_hl_2
sigur_hl
tbtest1
transit_hla
ya_control_v1


Sneaky Ema edits/merges/ports/recoveries:
------------------
adv_office_robbery
agpain_rb
ahl_blockwar
aiglass_hl
ak_rh_warehouse_ns
arc-novus-4
archivesp01
army
as_crazytank as_crazytank2
assault
base_attack1 base_attack2
battlefield_rush2
canal-r
cart_attack
casarosada_r_old
conc_speed
conc_speed2
conc_train
concmap_esadvr-rb
concmap_tg
conundrum2_1 conundrum2_2 conundrum2_3 conundrum2_4
cs_snowy
cs_zoption
d_trick1_hl
dccoop1_hl
deathrun_autobaw
deltapreview
devilmesa
die2_1 die2_2
die_way
dmc_dm1
dod_gauntlet
dod_smallbridge
e1m1_sc dmc_e1m2 q1_e1m3 dmc_e1m5_rb dmc_e1m8_rb dmc_dm1 e2m2 dmc_e3m3_rb dmc_dm5 e4m3 dmc_end pse_3_rb
eod_0_rb eod_1_rb eod_2_rb eod_3_rb
esad_raceconc2
etc2_1 etc2_2 etc2_3 etc2_4 etc2_5 etc2_6 etc2_7 etc2_8 etc2_9
expermnt_r
explosion
ey_compo
facilityfinal
forum
funhouse
fy_vertstadt
goodoldtime
halfmilegorge
hellbound_rb
hispeed
hl2_techdemo2_rb
hlywood
horde_short
hq_rb
id-ow
inf_isc
intrvw1
isp2 isp3 isp4 isp5 isp6
ispitatel
ivy
kune_rb-hl (merge of kune maps + 1 nipper map)
kz_6fd_volcano
kz_hauntedhouse_v2
kz_lain
kz_wtc_v2
lbm1
lowammo
ms2_starrancor
of_office
optimumfear_rb-r
overrun
pac_man
pve_building_rb
quad_2v2
renegade_r
rh_quarters
rjumptfc
roadofdestiny_rb
savesven_v1
saving_private_tower
sb3
sc_ature sc_ature2 sc_ature3 sc_ature3a sc_ature4 sc_ature5 sc_ature6 sc_ature7 sc_ature8_rb
sc_braindrain_beta3
sc_cp3
sc_soiled_b1
scramble
secretgarden
sigur sigur_2 sigur_3
snwconc_r
splcr3
stacja
stockade
strafe_porn
temp-916-r
the_stairs
torn2
tsm_r_v3
walkingsimulator_isc_hlb_1 walkingsimulator_isc_hl_2 walkingsimulator_isc_hla_3 walkingsimulator_isc_hla_4
xen_escape1a
xenomorphic



Many of these maps are not released on scmapdb, and are out of scope for v1 of the mod, but still appreciated.
Skill maps (kz_) and horde maps are strictly limited until I have a better system to prevent them being played too often.
```
  
</details>
